File: js/linkedList.js
"use strict";

class Node {
    constructor(data, next=null){
        this.data = data;
        this.next = next;
    }
}

class LinkedList {
    constructor(iterable=null) {
        this.head = null;
        this._counter = 0;
        if(Array.isArray(iterable)){
            iterable.map(x => this.push(x));
        }
    }

    push(val){
        let newHead = new Node(val, this.head);
        this.head = newHead;
        this._counter ++;
    }

    pop(){
        if(this.head === null){
            throw 'IndexError: Nothing to pop.';
        }
        let output = this.head;
        this.head = this.head.next;
        this._counter --;
        return output.data;
    }

    search(val){
        let searchThrough = this.head;
        try{
            while(this.head){
                if(val === searchThrough.data){
                    return searchThrough.data;
                }else{
                    searchThrough = searchThrough.next;
                }
            }
        }catch(e){
            return 'No such Node';
        }
    }

    remove(val){
        let currentNode = this.head;
        let previousNode = null;
        let found = false;
        if(currentNode === null){
            throw 'IndexError: Nothing in the list.';
        }
        try{
            while(currentNode && found === false){
                if(val === currentNode.data){
                    found = true;
                }else{
                    previousNode = currentNode;
                    currentNode = currentNode.next;
                }
            }
            if(previousNode === null){
                this.pop();
            }else if(currentNode.next === null){
                previousNode.next = null;
            }else{
                previousNode.next = currentNode.next;
            }
        }catch(e){
            throw 'ValueError: No such Node.';
        }
        this._counter --;
    }

    display(){
        let node = this.head;
        let displayThis = [];
        while(node){
            displayThis.push(node.data);
            node = node.next;
        }
        return displayThis;
    }

    size(){
        return this._counter;
    }
}

if(require.main === module){
    var list = new LinkedList([1, 2, 3, 4, 5]);
    console.log(list.display());
    console.log(list.size());
}

File: src/linked_list.py
"""A Linked List object."""


class Node(object):
    """A node object."""

    def __init__(self, data, nxt=None):
        """The init values of a Node object."""
        self.data = data
        self.nxt = nxt


class LinkedList(object):
    """A Linked List object."""

    def __init__(self, iterable=None):
        """The init values of a Linked List object."""
        self.head = None
        self._counter = 0
        if isinstance(iterable, (str, list, tuple)):
            for item in iterable:
                self.push(item)

    def push(self, val):
        """Add a new Node containing the given value."""
        new_head = Node(val, nxt=self.head)
        self.head = new_head
        self._counter += 1

    def pop(self):
        """Remove and return the head Node and return it's value."""
        if not self.head:
            raise IndexError("The list is empty.")
        output = self.head.data
        self.head = self.head.nxt
        self._counter -= 1
        return output

    def size(self):
        """Return the size of the Linked List."""
        return self._counter

    def search(self, val):
        """Return the Node containing the given value."""
        curr = self.head
        while curr:
            if val == curr.data:
                return curr
            curr = curr.nxt
        return None

    def remove(self, node):
        """Remove the Node containing the given value."""
        if not self.head:
            raise IndexError("The list is empty.")
        if self.head == node:
            self.head = self.head.nxt
            self._counter -= 1
            return
        curr = self.head
        while curr.nxt:
            if curr.nxt == node:
                curr.nxt = node.nxt
                self._counter -= 1
                return
            curr = curr.nxt

    def display(self):
        """Display the Linked List."""
        curr = self.head
        output = []
        while curr:
            output.append(curr.data)
            curr = curr.nxt
        return tuple(output).__str__()

    def __len__(self):
        """Overwrite default return for the len function."""
        return self._counter

    def __str__(self):
        """Overwrite default return for the print function."""
        return self.display()
